spring-integration-webflux/src/main/java/org/springframework/integration/webflux/inbound/WebFluxInboundEndpoint.java

```diff
@@ -151,16 +152,20 @@ private Mono<Void> doHandle(ServerWebExchange exchange) {
 						new RequestEntity<>(body, exchange.getRequest().getHeaders(),
 								exchange.getRequest().getMethod(), exchange.getRequest().getURI()))
 				.flatMap(entity -> buildMessage(entity, exchange))
-				.flatMap(requestTuple -> {
-					if (isExpectReply()) {
-						return sendAndReceiveMessageReactive(requestTuple.getT1())
-								.flatMap(replyMessage -> populateResponse(exchange, replyMessage));
-					}
-					else {
-						send(requestTuple.getT1());
-						return setStatusCode(exchange, requestTuple.getT2());
-					}
-				})
+				.flatMap(requestTuple ->
+						Mono.deferContextual(contextView -> {
+							if (isExpectReply()) {
+								return sendAndReceiveMessageReactive(requestTuple.getT1())
+										.flatMap(replyMessage -> populateResponse(exchange, replyMessage));
+							}
+							else {
+								var scope = ContextSnapshot.setAllThreadLocalsFrom(contextView);
```

Comment:
Why is this not required in the send/receive case?

Reply:
It is covered by the `MessagingGatewaySupport` change in this PR. 
We need such an action only when we lose a reactive scope.
This `send()` returns nothing, but `sendAndReceiveMessageReactive()` returns a `Mono` where respective context is still present. Therefore we have to repeat a set to `ThreadLocal` over there where we lose the reactive stream context again.
That's why I have those separate tests in this PR.

NOTE: The RSocket support will work as well just because we have a fix in the `MessagingGatewaySupport`  and it does not use a plain `send()`.